MRG: improve feeds

jasmainak opened the pull request:
Giving this another shot. To complement efforts by @cjayb

jasmainak commented on hnn_core/feed.py:
```diff
@@ -121,10 +121,7 @@ def set_event_times(self):
 
         all_syn_weights_zero = True
         for key in self.params.keys():
-            if key.startswith('L2Pyr') or \
```
@rythorpe this commit is for you :)

jasmainak commented:
okay this PR is beginning to get big. I am now planning to add unit tests for each of the different functions I documented. Might get to it next week. But in the meanwhile, would appreciate another set of eyes to look at it (@cjayb @rythorpe @ntolley ?). 

The easiest way to review would be to look at the [new feed.py file](https://github.com/jasmainak/hnn-core/blob/improve_feeds/hnn_core/feed.py) along with the diff. Here is a summary  of the main changes:

- simplification of seed generation with a `_get_prng` method.
- dismantling of `ExtFeed` class to `feed_event_times` function.
- unifying truncating event_times > 0 and sorting of event_times in one function

I hope this helps a lot in seeing clearly what we should do next with feeds.

cjayb commented on hnn_core/feed.py:
```diff
@@ -35,252 +74,238 @@ class ExtFeed(object):
+        The events per cycle. Must be 1 or 2.
+    prng : instance of RandomState
+        The random state.
+    prng : instance of RandomState
```
prng2

cjayb commented on hnn_core/feed.py:
```diff
@@ -35,252 +74,238 @@ class ExtFeed(object):
+
+    t_input = np.array([])
+    if events_per_cycle == 2:
+        # Two arrays store doublet times
+        t_input_low = t_array - 5
+        t_input_high = t_array + 5
+        # Array with ALL stimulus times for input
+        # np.append concatenates two np arrays
+        t_input = np.append(t_input_low, t_input_high)
```
So this isn't a comment on this PR, more to the code: Could we get one of the OPs to give us a link to the part of documentation where this is specified?

jasmainak replied:
maybe @rythorpe or @ntolley can help since they use the GUI?

rythorpe replied:
I don't understand. What specifically are you looking for here @cjayb?

cjayb replied:
The code is now simplified to

```python
return np.append(t_array - 5, t_array + 5)
```

but doesn't it creep you out that there are undocumented, hard-coded time shifts there? Clearly it's creating spike "doublets", and certainly for a very specific reason. I was just wondering whether anyone still remembers what the reason was?

jasmainak replied:
> but doesn't it creep you out that there are undocumented, hard-coded time shifts there?

LOL yes :) Maybe it makes sense for some reason, I don't know ...

I think @rythorpe or @ntolley can throw some light on what this function does and how the spike train is supposed to look like and why. I opened the GUI many aeons ago ...

ntolley replied:
Perhaps you should restore the comment explaining how this is equivalent to the spikes/burst parameter in GUI?. As for the fixed time shifts, I'm guessing there is a biologic justification for a doublet with a 10 ms ISI. 

I see that it was added in a pretty large commit in the [GUI](https://github.com/jonescompneurolab/hnn/commit/2161e2c75bf0378080848093b987f7424b1c6f85#diff-84da9608764aa67ace81318363636bd272c6053c686b2be6616187c04325bcfdR170).

rythorpe replied:
I think the 10 ms is a somewhat standard approximation for the amount of time between spikes within a thalamic burst. The [HNN webset/tutorials](https://jonescompneurolab.github.io/hnn-under_the_hood/04_evoked-rhythmic-inputs/04_evoked-rhythmic-inputs) don't explicitly address this, however, they do reference [this](https://pubmed.ncbi.nlm.nih.gov/16061522/) paper that discusses the bursting properties of thalamic drive into cortex.

rythorpe replied:
Ideally we'll define an API in a later PR for creating rhythmic feeds that allows the user to define the intra-burst ISI.

jasmainak replied:
done, I updated the docstring :) A demo of the GUI for these functions would be nice one day, now that we have everything nicely documented in the code.

cjayb reviewed (approved):
Apart from a single typo, I find this to be only positive. A lot of the nestedness is gone, even though there remain a lot of idiosyncrasies from "HNN Classic". With #191 we can get all this into `Network`, where it'll be easier to think about a good API (and tests!) for `prng` as well. Definite +1 from me!

rythorpe commented on hnn_core/feed.py:
```diff
@@ -35,252 +74,238 @@ class ExtFeed(object):
+        zero_ampa_nmda = (params[target_cell_type][0] <= 0.0 and
+                          params[target_cell_type][1] <= 0.0)
+
+    all_syn_weights_zero = True
+    for key in params.keys():
+        if key.startswith(('L2Pyr', 'L5Pyr', 'L2Bask', 'L5Bask')):
+            if params[key][0] > 0.0:
+                all_syn_weights_zero = False
```
+1 for this!

rythorpe commented on hnn_core/feed.py:
```diff
@@ -35,252 +74,238 @@ class ExtFeed(object):
+    # Return values not checked: False if all weights for given feed type
+    # are zero. Designed to be silent so that zeroing input weights
+    # effectively disables each.
+    zero_ampa_nmda = False
```
This is nitpicky, but could we rename `zero_ampa_nmda` to `target_syn_weights_zero` just to make it consistent with `all_sysn_weights_zero`? This way it's super clear which one applies to unique feeds and which one applies to common feeds.

rythorpe commented on hnn_core/feed.py:
```diff
@@ -35,252 +74,238 @@ class ExtFeed(object):
+            prng=prng,
+            prng2=prng2)
+
+    # brute force remove non-zero times. Might result in fewer vals
+    # than desired
+    # values MUST be sorted for VecStim()!
+    if len(event_times) > 0:
+        event_times = event_times[event_times > 0]
+        event_times.sort()
+        event_times.tolist()
```
Why is this necessary? Shouldn't this be enforced when `event_times` is created?

jasmainak replied:
The numpy functions return arrays, so you need to convert them to a list

rythorpe replied:
My question was more about why `event_times > 0` isn't enforced from the beginning, but I think I get it now.

jasmainak replied:
Actually it's a very good point now that you mention it. Only the poisson inputs are guaranteed to have `event_times > 0`. For the rest, it depends. For example, if it's Gaussian and the mean is too close to 0., then some events in the beginning will be truncated. Honestly, I don't get the logic of that. If you remove some events, then what you get is actually a [truncated Gaussian distribution](https://en.wikipedia.org/wiki/Truncated_normal_distribution) not a Gaussian distribution.

rythorpe commented on hnn_core/feed.py:
```diff
@@ -7,8 +7,47 @@
+
+    Returns
+    -------
+    prng : instance of RandomState
+        The seed for events assuming a given start time.
+    prng2 : instance of RandomState
+        The seed for start times.
```
I'm a little confused on the use of these two prng values. Could we clarify them a little in the documentation?

jasmainak replied:
to be honest, I am a bit confused myself. I thought `prng2` is used to get the start time jitter but actually it turns out if `t0 = -1` then `prng` is used. See: https://github.com/jonescompneurolab/hnn-core/blob/56f3565dcee8592872ac79a911b250c830d1d67b/hnn_core/feed.py#L225-L230

Nonetheless I clarified in the documentation assuming the intent was actually for the start time for both cases.

rythorpe replied:
It looks like `prng2` is only used for rhythmic inputs and defines the jittered start time of the rhythmic burst. Then, `prng` is used to define the jittered input of single input within that burst. If my understanding is correct, we may want to simplify `_get_prng()` by elimnating any reference to `prng2` and then adding the logic for creating a second random state variable to `feed_event_times()` only if a rhythmic (i.e., common) input is in the making.

rythorpe replied:
Actually, this will probably get resolved when we work on a more flexible API for creating new feeds so let's ignore this for now.

jasmainak replied:
That makes sense. Note though the current "API". Every function has the last argument as `prng`. This is how it should be. The user should be able to supply their own `prng`. It shouldn't be generated opaquely inside a function or class

rythorpe commented on hnn_core/feed.py:
```diff
@@ -7,8 +7,47 @@
+    return prng, prng2
+
+
+def feed_event_times(feed_type, target_cell_type, params, gid):
```
Would this be a good time to start explicitly defining the elements of the `params` argument for this function?

jasmainak replied:
not sure if it would be a good use of our time to document what is created as an output of `create_pext`. I think it _might_ be helpful to move `create_pext` into `feed.py` so folks can see what parameter corresponds to what. We would avoid duplicating what is in `params.py` just for documentation ... Let me know if you want me to do that or if you'd rather keep that in `params.py`

cjayb replied:
I'm a strong +1 on leaving as-is for this PR. create_pext is a symptom of unwieldy parameter handling, so fixing (or documenting) one or two methods, I think, has lower priority than figuring out what to do about the root cause.

rythorpe replied:
We can wait on this. At some point we need to start dismantling the large param dicts that get passed around and I'm beginning to wonder if the best way to start this process is to begin explicitly defining function arguments where we can.

cjayb replied:
Very likely so, let's see what we can do with a new feeds (drives) API

jasmainak replied:
+100 on explicitly defining function arguments where we can! It's the best way to avoid opaque code

jasmainak commented:
Alright, I'm done here.  Addressed all the comments and updated tests.

`feed.py` is massively simplified. Half of the file is now just docstrings. I don't quite get `_create_common_input` is supposed to do so I didn't test any of the functionality other than checking that errors are raised correctly. I'd leave this checking for another PR. But any help from GUI users to explain the purpose of this would be very helpful :)

The poisson input function is simplified and it uses `numpy` now. 

Before you decide to merge, please make sure that the example plots look similar to what we see on the current version of the website. Thank a lot for reviewing guys!

jasmainak commented:
PR rebased :)

cjayb commented:
Sweet! I'm offline until tomorrow so feel free to merge if you think it's clean enough.

cjayb commented:
Bringing discussion on `net.trial_event_times` here @jasmainak :

> It should be dict (keys are the src_type) of list (gids) of list (n_trials) of list (n_events) ? I think it will help unify things related to trial jitter etc.

The place look at is `NetworkBuilder._create_cells_and_feeds:L404`: the for-loop over all gids on this rank.  I suppose you could do:

```python
        for gid in self._gid_list:  # L404
            src_type, src_pos, is_cell = self.net._get_src_type_and_pos(gid)
            this_src_event_times = self.net.trial_event_times[src_type]
    ...
        else:
            gid_idx = gid - self.net.gid_ranges[src_type][0]  # still needed
            et = this_src_event_times[self.trial_idx][gid_idx]
...
```

Want to try to flip it around (to dict of lists)? Perhaps rename  `net.trial_event_times` (back) to `net.feed_event_times`. NB for single-trial simulations this means that the first list dimension `len` is `1`.

cjayb commented:
I'm also +1 for a merge.

jasmainak commented on hnn_core/feed.py:
```diff
@@ -36,251 +70,207 @@ class ExtFeed(object):
-        if lamtha > 0.:
-            t_gen = t0 + self._t_wait(lamtha)
-            if t_gen < T:
-                np.append(val_pois, t_gen)
```
@rythorpe found the source of the discrepancy you noticed in the gamma example. When you use [`np.append`](https://numpy.org/doc/stable/reference/generated/numpy.append.html), the return value needs to be `val_pois`. The array is not modified in-place like a  list. So, this is actually a bug in the old code. One spike is missed at the beginning. Should we consider this a bugfix or should I push a commit to revert to the old behavior?

rythorpe replied:
Oh I see. This is a good find. My vote is to move forward with your current solution as a bug fix and document it in `whats_new.rst`.

jasmainak commented on hnn_core/feed.py:
```diff
@@ -36,251 +70,207 @@ class ExtFeed(object):
+        raise ValueError(f'Rate must be > 0. Got {lamtha}')
+
+    event_times = list()
+    t_gen = t0
```
doing:

```py
t_gen = t0 + prng.exponential(1. / lamtha) * 1000.
```

will revert to old behavior @rythorpe

rythorpe replied:
Do need to address anything here @jasmainak or was this addressed above?

jasmainak replied:
I think we're good here

jasmainak commented:
> Want to try to flip it around (to dict of lists)?

Let's keep this for another PR. I don't want to increase the size of this PR any further :)

jasmainak commented:
I updated the `whats_new` with the bugfix note

rythorpe commented:
Any other last minute fixes or are we good to rebase and merge?

jasmainak commented:
Good to merge!

rythorpe commented:
Done! Thanks @jasmainak!
